From 2f9934c7722381c720078855fa9f71d8b025be6f Mon Sep 17 00:00:00 2001
From: Koding teammate richard <richard@ip-10-0-0-143.(none)>
Date: Mon, 4 Mar 2013 13:09:25 +0000
Subject: [PATCH] Allow mknod for userns root.

---
 fs/namei.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/fs/namei.c b/fs/namei.c
index 0f018b1..8955259 100644
--- a/fs/namei.c
+++ b/fs/namei.c
@@ -3136,7 +3136,7 @@ int vfs_mknod(struct inode *dir, struct dentry *dentry, umode_t mode, dev_t dev)
 	if (error)
 		return error;
 
-	if ((S_ISCHR(mode) || S_ISBLK(mode)) && !capable(CAP_MKNOD))
+	if ((S_ISCHR(mode) || S_ISBLK(mode)) && !nsown_capable(CAP_MKNOD))
 		return -EPERM;
 
 	if (!dir->i_op->mknod)
-- 
1.7.10.4

