From 6ce604e4de51bdcce72bf59a78121ea9c8eb08ec Mon Sep 17 00:00:00 2001
From: Richard <richard@koding.com>
Date: Fri, 14 Dec 2012 03:18:05 +0000
Subject: [PATCH] Allow multiple writes to uid_map/gid_map.

---
 kernel/user_namespace.c |    8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/kernel/user_namespace.c b/kernel/user_namespace.c
index dafa125..1fed8e9 100644
--- a/kernel/user_namespace.c
+++ b/kernel/user_namespace.c
@@ -561,9 +561,6 @@ static ssize_t map_write(struct file *file, const char __user *buf,
 	mutex_lock(&id_map_mutex);
 
 	ret = -EPERM;
-	/* Only allow one successful write to the map */
-	if (map->nr_extents != 0)
-		goto out;
 
 	/* Require the appropriate privilege CAP_SETUID or CAP_SETGID
 	 * over the user namespace in order to set the id mapping.
@@ -592,7 +589,10 @@ static ssize_t map_write(struct file *file, const char __user *buf,
 	/* Parse the user data */
 	ret = -EINVAL;
 	pos = kbuf;
-	new_map.nr_extents = 0;
+        memcpy(new_map.extent, map->extent,
+                map->nr_extents*sizeof(map->extent[0]));
+        smp_wmb();
+        new_map.nr_extents = map->nr_extents;
 	for (;pos; pos = next_line) {
 		extent = &new_map.extent[new_map.nr_extents];
 
-- 
1.7.10.4

