// log4js/build/stage/lib/appenders/hookio.js
/* automatically generated by JSCoverage - do not edit */
if (typeof _$jscoverage === 'undefined') _$jscoverage = {};
if (! _$jscoverage['appenders/hookio.js']) {
  _$jscoverage['appenders/hookio.js'] = [];
  _$jscoverage['appenders/hookio.js'][1] = 0;
  _$jscoverage['appenders/hookio.js'][2] = 0;
  _$jscoverage['appenders/hookio.js'][3] = 0;
  _$jscoverage['appenders/hookio.js'][4] = 0;
  _$jscoverage['appenders/hookio.js'][6] = 0;
  _$jscoverage['appenders/hookio.js'][7] = 0;
  _$jscoverage['appenders/hookio.js'][8] = 0;
  _$jscoverage['appenders/hookio.js'][9] = 0;
  _$jscoverage['appenders/hookio.js'][10] = 0;
  _$jscoverage['appenders/hookio.js'][11] = 0;
  _$jscoverage['appenders/hookio.js'][12] = 0;
  _$jscoverage['appenders/hookio.js'][13] = 0;
  _$jscoverage['appenders/hookio.js'][17] = 0;
  _$jscoverage['appenders/hookio.js'][19] = 0;
  _$jscoverage['appenders/hookio.js'][20] = 0;
  _$jscoverage['appenders/hookio.js'][21] = 0;
  _$jscoverage['appenders/hookio.js'][22] = 0;
  _$jscoverage['appenders/hookio.js'][26] = 0;
  _$jscoverage['appenders/hookio.js'][27] = 0;
  _$jscoverage['appenders/hookio.js'][28] = 0;
  _$jscoverage['appenders/hookio.js'][30] = 0;
  _$jscoverage['appenders/hookio.js'][33] = 0;
  _$jscoverage['appenders/hookio.js'][35] = 0;
  _$jscoverage['appenders/hookio.js'][36] = 0;
  _$jscoverage['appenders/hookio.js'][39] = 0;
  _$jscoverage['appenders/hookio.js'][40] = 0;
  _$jscoverage['appenders/hookio.js'][41] = 0;
  _$jscoverage['appenders/hookio.js'][42] = 0;
  _$jscoverage['appenders/hookio.js'][43] = 0;
  _$jscoverage['appenders/hookio.js'][44] = 0;
  _$jscoverage['appenders/hookio.js'][46] = 0;
  _$jscoverage['appenders/hookio.js'][49] = 0;
  _$jscoverage['appenders/hookio.js'][50] = 0;
  _$jscoverage['appenders/hookio.js'][51] = 0;
  _$jscoverage['appenders/hookio.js'][53] = 0;
  _$jscoverage['appenders/hookio.js'][58] = 0;
  _$jscoverage['appenders/hookio.js'][59] = 0;
  _$jscoverage['appenders/hookio.js'][60] = 0;
  _$jscoverage['appenders/hookio.js'][61] = 0;
  _$jscoverage['appenders/hookio.js'][64] = 0;
  _$jscoverage['appenders/hookio.js'][65] = 0;
  _$jscoverage['appenders/hookio.js'][66] = 0;
  _$jscoverage['appenders/hookio.js'][67] = 0;
  _$jscoverage['appenders/hookio.js'][68] = 0;
  _$jscoverage['appenders/hookio.js'][69] = 0;
  _$jscoverage['appenders/hookio.js'][71] = 0;
  _$jscoverage['appenders/hookio.js'][74] = 0;
  _$jscoverage['appenders/hookio.js'][75] = 0;
  _$jscoverage['appenders/hookio.js'][76] = 0;
}
_$jscoverage['appenders/hookio.js'][1]++;
var log4js = require("../log4js");
_$jscoverage['appenders/hookio.js'][2]++;
var layouts = require("../layouts");
_$jscoverage['appenders/hookio.js'][3]++;
var Hook = require("hook.io").Hook;
_$jscoverage['appenders/hookio.js'][4]++;
var util = require("util");
_$jscoverage['appenders/hookio.js'][6]++;
var Logger = (function createLogger(options) {
  _$jscoverage['appenders/hookio.js'][7]++;
  var self = this;
  _$jscoverage['appenders/hookio.js'][8]++;
  var actualAppender = options.actualAppender;
  _$jscoverage['appenders/hookio.js'][9]++;
  Hook.call(self, options);
  _$jscoverage['appenders/hookio.js'][10]++;
  self.on("hook::ready", (function hookReady() {
  _$jscoverage['appenders/hookio.js'][11]++;
  self.on("*::" + options.name + "::log", (function log(loggingEvent) {
  _$jscoverage['appenders/hookio.js'][12]++;
  deserializeLoggingEvent(loggingEvent);
  _$jscoverage['appenders/hookio.js'][13]++;
  actualAppender(loggingEvent);
}));
}));
});
_$jscoverage['appenders/hookio.js'][17]++;
util.inherits(Logger, Hook);
_$jscoverage['appenders/hookio.js'][19]++;
function deserializeLoggingEvent(loggingEvent) {
  _$jscoverage['appenders/hookio.js'][20]++;
  loggingEvent.startTime = new Date(loggingEvent.startTime);
  _$jscoverage['appenders/hookio.js'][21]++;
  loggingEvent.level.toString = (function levelToString() {
  _$jscoverage['appenders/hookio.js'][22]++;
  return loggingEvent.level.levelStr;
});
}
_$jscoverage['appenders/hookio.js'][26]++;
function initHook(hookioOptions) {
  _$jscoverage['appenders/hookio.js'][27]++;
  var loggerHook;
  _$jscoverage['appenders/hookio.js'][28]++;
  if (hookioOptions.mode === "master") {
    _$jscoverage['appenders/hookio.js'][30]++;
    loggerHook = new Logger(hookioOptions);
  }
  else {
    _$jscoverage['appenders/hookio.js'][33]++;
    loggerHook = new Hook(hookioOptions);
  }
  _$jscoverage['appenders/hookio.js'][35]++;
  loggerHook.start();
  _$jscoverage['appenders/hookio.js'][36]++;
  return loggerHook;
}
_$jscoverage['appenders/hookio.js'][39]++;
function getBufferedHook(hook, eventName) {
  _$jscoverage['appenders/hookio.js'][40]++;
  var hookBuffer = [];
  _$jscoverage['appenders/hookio.js'][41]++;
  var hookReady = false;
  _$jscoverage['appenders/hookio.js'][42]++;
  hook.on("hook::ready", (function emptyBuffer() {
  _$jscoverage['appenders/hookio.js'][43]++;
  hookBuffer.forEach((function logBufferItem(loggingEvent) {
  _$jscoverage['appenders/hookio.js'][44]++;
  hook.emit(eventName, loggingEvent);
}));
  _$jscoverage['appenders/hookio.js'][46]++;
  hookReady = true;
}));
  _$jscoverage['appenders/hookio.js'][49]++;
  return (function log(loggingEvent) {
  _$jscoverage['appenders/hookio.js'][50]++;
  if (hookReady) {
    _$jscoverage['appenders/hookio.js'][51]++;
    hook.emit(eventName, loggingEvent);
  }
  else {
    _$jscoverage['appenders/hookio.js'][53]++;
    hookBuffer.push(loggingEvent);
  }
});
}
_$jscoverage['appenders/hookio.js'][58]++;
function createAppender(hookioOptions) {
  _$jscoverage['appenders/hookio.js'][59]++;
  var loggerHook = initHook(hookioOptions);
  _$jscoverage['appenders/hookio.js'][60]++;
  var loggerEvent = hookioOptions.name + "::log";
  _$jscoverage['appenders/hookio.js'][61]++;
  return getBufferedHook(loggerHook, loggerEvent);
}
_$jscoverage['appenders/hookio.js'][64]++;
function configure(config) {
  _$jscoverage['appenders/hookio.js'][65]++;
  var actualAppender;
  _$jscoverage['appenders/hookio.js'][66]++;
  if (config.appender && config.mode === "master") {
    _$jscoverage['appenders/hookio.js'][67]++;
    log4js.loadAppender(config.appender.type);
    _$jscoverage['appenders/hookio.js'][68]++;
    actualAppender = log4js.appenderMakers[config.appender.type](config.appender);
    _$jscoverage['appenders/hookio.js'][69]++;
    config.actualAppender = actualAppender;
  }
  _$jscoverage['appenders/hookio.js'][71]++;
  return createAppender(config);
}
_$jscoverage['appenders/hookio.js'][74]++;
exports.name = "hookio";
_$jscoverage['appenders/hookio.js'][75]++;
exports.appender = createAppender;
_$jscoverage['appenders/hookio.js'][76]++;
exports.configure = configure;
_$jscoverage['appenders/hookio.js'].source = ["var log4js = require('../log4js');","var layouts = require('../layouts');","var Hook = require('hook.io').Hook;","var util = require('util');","","var Logger = function createLogger(options) {","  var self = this;","  var actualAppender = options.actualAppender;","  Hook.call(self, options);","  self.on('hook::ready', function hookReady() {","    self.on('*::' + options.name + '::log', function log(loggingEvent) {","      deserializeLoggingEvent(loggingEvent);","      actualAppender(loggingEvent);","    });","  });","}","util.inherits(Logger, Hook);","","function deserializeLoggingEvent(loggingEvent) {","  loggingEvent.startTime = new Date(loggingEvent.startTime);","  loggingEvent.level.toString = function levelToString() {","    return loggingEvent.level.levelStr;","  };","}","","function initHook(hookioOptions) {","  var loggerHook;","  if (hookioOptions.mode === 'master') {","    // Start the master hook, handling the actual logging","    loggerHook = new Logger(hookioOptions);","  } else {","    // Start a worker, just emitting events for a master","    loggerHook = new Hook(hookioOptions);","  }","  loggerHook.start();","  return loggerHook;","}","","function getBufferedHook(hook, eventName) {","  var hookBuffer = [];","  var hookReady = false;","  hook.on('hook::ready', function emptyBuffer() {","    hookBuffer.forEach(function logBufferItem(loggingEvent) {","      hook.emit(eventName, loggingEvent);","    })","    hookReady = true;","  });","","  return function log(loggingEvent) {","    if (hookReady) {","      hook.emit(eventName, loggingEvent);","    } else {","      hookBuffer.push(loggingEvent);","    }","  }","}","","function createAppender(hookioOptions) {","  var loggerHook = initHook(hookioOptions);","  var loggerEvent = hookioOptions.name + '::log';","  return getBufferedHook(loggerHook, loggerEvent);","}","","function configure(config) {","  var actualAppender;","  if (config.appender &amp;&amp; config.mode === 'master') {","    log4js.loadAppender(config.appender.type);","    actualAppender = log4js.appenderMakers[config.appender.type](config.appender);","    config.actualAppender = actualAppender;","  }","  return createAppender(config);","}","","exports.name = 'hookio';","exports.appender = createAppender;","exports.configure = configure;"];
